Add `codeserver-python` rock and tests

BON4 opened the pull request:
### Changes
---
Added Codeserver & Jupyter rock  v1.10.0-rc.1 and rock test with codeserver access test

### Testing
---
```
> tox -vve sanity

============================================================================================= test session starts ==============================================================================================
platform linux -- Python 3.9.21, pytest-8.3.5, pluggy-1.5.0 -- /home/ubuntu/kubeflow-rocks/codeserver-python/.tox/sanity/bin/python
cachedir: .tox/sanity/.pytest_cache
rootdir: /home/ubuntu/kubeflow-rocks/codeserver-python
plugins: anyio-4.9.0
collected 1 item                                                                                                                                                                                               

tests/test_rock.py::test_rock PASSED                                                                                                                                                                     [100%]

=============================================================================================== warnings summary ===============================================================================================
tests/test_rock.py:10
  /home/ubuntu/kubeflow-rocks/codeserver-python/tests/test_rock.py:10: PytestUnknownMarkWarning: Unknown pytest.mark.abort_on_fail - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.abort_on_fail

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================================================================= 1 passed, 1 warning in 1.48s =========================================================================================
sanity: 2548 I exit 0 (2.05 seconds) /home/ubuntu/kubeflow-rocks/codeserver-python> pytest -v --tb native --show-capture=all --log-cli-level=INFO /home/ubuntu/kubeflow-rocks/codeserver-python/tests pid=51109 [tox/execute/api.py:294]
sanity: 2548 W commands[1]> python /home/ubuntu/kubeflow-rocks/codeserver-python/tests/test_access.py [tox/tox_env/api.py:434]
d73df3c96b3c
d73df3c96b3c
sanity: 6479 I exit 0 (3.93 seconds) /home/ubuntu/kubeflow-rocks/codeserver-python> python /home/ubuntu/kubeflow-rocks/codeserver-python/tests/test_access.py pid=51407 [tox/execute/api.py:294]
  sanity: OK (6.07=setup[0.09]+cmd[2.05,3.93] seconds)
  congratulations :) (6.21 seconds)


```

mvlassis commented on codeserver-python/rockcraft.yaml:
```diff
@@ -0,0 +1,246 @@
+    stage-packages:
+      - python3-distutils
+    override-build: |
+      set -xe
```
suggestion: This is unneeded, as the build will fail if there is a command with a non-zero exit code
```suggestion
```

BON4 replied:
Done

mvlassis commented on codeserver-python/rockcraft.yaml:
```diff
@@ -0,0 +1,246 @@
+      echo ". /opt/conda/etc/profile.d/conda.sh" >> "${HOME}/.bashrc"
+      echo ". /opt/conda/etc/profile.d/conda.sh" >> "${CRAFT_PART_INSTALL}/etc/profile"
+      echo "conda activate base" >> "${HOME}/.bashrc"
+            echo "conda activate base" >> "${CRAFT_PART_INSTALL}/etc/profile"
```
```suggestion
        echo "conda activate base" >> "${CRAFT_PART_INSTALL}/etc/profile"
```

mvlassis commented on codeserver-python/rockcraft.yaml:
```diff
@@ -0,0 +1,246 @@
+      chmod 2775 ${CRAFT_PART_INSTALL}/${CONDA_DIR}
+      echo ". /opt/conda/etc/profile.d/conda.sh" >> "${HOME}/.bashrc"
+      echo ". /opt/conda/etc/profile.d/conda.sh" >> "${CRAFT_PART_INSTALL}/etc/profile"
+      echo "conda activate base" >> "${HOME}/.bashrc"
+            echo "conda activate base" >> "${CRAFT_PART_INSTALL}/etc/profile"
```
```suggestion
      echo "conda activate base" >> "${HOME}/.bashrc"
      echo "conda activate base" >> "${CRAFT_PART_INSTALL}/etc/profile"
```

BON4 replied:
Done

mvlassis commented on codeserver-python/rockcraft.yaml:
```diff
@@ -0,0 +1,246 @@
+#Based on the following Dockerfiles:
```
```suggestion
# Based on the following Dockerfiles:
```

BON4 replied:
Done

mvlassis commented on codeserver-python/rockcraft.yaml:
```diff
@@ -0,0 +1,246 @@
+  in a user namespace, and enables users to perform CPU-accelerated tasks
+  using PyTorch from within a Jupyter Notebook.
+
+  Both PyTorch and Jupyter are installed in Conda environment, which is
```
```suggestion
  Both PyTorch and Jupyter are installed in a Conda environment, which is
```

BON4 replied:
Done

mvlassis commented on codeserver-python/rockcraft.yaml:
```diff
@@ -0,0 +1,246 @@
+#Based on the following Dockerfiles:
+# https://github.com/kubeflow/kubeflow/blob/v1.10.0-rc.1/components/example-notebook-servers/base/Dockerfile
+# https://github.com/kubeflow/kubeflow/blob/v1.10.0-rc.1/components/example-notebook-servers/codeserver/Dockerfile
+# https://github.com/kubeflow/kubeflow/blob/v1.10.0-rc.1/components/example-notebook-servers/codeserver-python/Dockerfile
```
suggestion(blocking): Let's update the tag as well since `v1.10.0` has been released
```suggestion
# https://github.com/kubeflow/kubeflow/blob/v1.10.0/components/example-notebook-servers/base/Dockerfile
# https://github.com/kubeflow/kubeflow/blob/v1.10.0/components/example-notebook-servers/codeserver/Dockerfile
# https://github.com/kubeflow/kubeflow/blob/v1.10.0/components/example-notebook-servers/codeserver-python/Dockerfile
```

BON4 replied:
Done

mvlassis commented on codeserver-python/rockcraft.yaml:
```diff
@@ -0,0 +1,246 @@
+
+  Both PyTorch and Jupyter are installed in Conda environment, which is
+  automatically activated.
+version: "1.10.0-rc.1"
```
suggestion(blocking): Let's update the tag
```suggestion
version: "1.10.0"
```

BON4 replied:
Done

mvlassis commented on codeserver-python/rockcraft.yaml:
```diff
@@ -0,0 +1,246 @@
+    plugin: nil
+    source: https://github.com/kubeflow/kubeflow
+    source-type: git
+    source-tag: v1.10.0-rc.1
```
suggestion(blocking): Let's update the tag
```suggestion
    source-tag: v1.10.0
```

mvlassis reviewed (commented):
Hey @BON4! Nice work, I made some small suggestions and will be testing the rock promptly

mvlassis commented on codeserver-python/tests/test_rock.py:
```diff
@@ -0,0 +1,24 @@
+    assert rock_services["codeserver-jupyter"]["startup"] == "enabled"
+
+    # verify that artifacts are in correct locations
+    subprocess.run(["docker", "run", LOCAL_ROCK_IMAGE, "exec", "ls", "-ls", "/opt/conda/bin/jupyter"], check=True)
```
Maybe we could add a test to ensure the `/usr/bin/code-server` executable also exists
```suggestion
    subprocess.run(["docker", "run", LOCAL_ROCK_IMAGE, "exec", "ls", "-ls", "/usr/bin/code-server], check=True)
    subprocess.run(["docker", "run", LOCAL_ROCK_IMAGE, "exec", "ls", "-ls", "/opt/conda/bin/jupyter"], check=True)
```

BON4 replied:
Done

mvlassis commented on codeserver-python/tox.ini:
```diff
@@ -0,0 +1,58 @@
+passenv = *
+allowlist_externals =
+    bash
+    skopeo
```
```suggestion
    rockcraft
```

BON4 replied:
Done

mvlassis commented on codeserver-python/tox.ini:
```diff
@@ -0,0 +1,58 @@
+             ROCK="$\{NAME\}_$\{VERSION\}_$\{ARCH\}.rock" && \
+             DOCKER_IMAGE=$NAME:$VERSION && \
+             echo "Exporting $ROCK to docker as $DOCKER_IMAGE" && \
+             skopeo --insecure-policy copy oci-archive:$ROCK docker-daemon:$DOCKER_IMAGE'
```
```suggestion
             rockcraft.skopeo --insecure-policy copy oci-archive:$ROCK docker-daemon:$DOCKER_IMAGE'
```

BON4 replied:
Done

mvlassis commented on codeserver-python/rockcraft.yaml:
```diff
@@ -0,0 +1,246 @@
+
+      # install - ipykernel
+      # NOTE: we need this for jupyter codeserver extension to work
+      echo "ipykernel ==${IPYKERNEL_VERSION}" >> ${CONDA_DIR}/conda-meta/pinned
```
```suggestion
      echo "ipykernel==${IPYKERNEL_VERSION}" >> ${CONDA_DIR}/conda-meta/pinned
```

BON4 replied:
Done

mvlassis commented on codeserver-python/rockcraft.yaml:
```diff
@@ -0,0 +1,246 @@
+    source-type: git
+    source-tag: v1.10.0-rc.1
+    build-packages:
+      - lsb-release
+      - wget
```
We should add the `libxml2` package, otherwise I get the following error:
```
:: Error while loading conda entry point: conda-libmamba-solver (libxml2.so.2: cannot open shared object file: No such file or directory)
```
```suggestion
      - lsb-release
      - wget
```

BON4 replied:
This will not solve the issue. It works when `--solver=classic` is provided for `conda install ipykernel`.

mvlassis commented on codeserver-python/rockcraft.yaml:
```diff
@@ -0,0 +1,246 @@
+      cp -r /lib/code-server-${CODESERVER_VERSION} ${CRAFT_PART_INSTALL}/lib/code-server
+      ln -sf /lib/code-server/bin/code-server ${CRAFT_PART_INSTALL}/usr/bin/code-server
+      cp -r "${CONDA_DIR}" "${CRAFT_PART_INSTALL}/opt"      
+      cp -r "${HOME}" "${CRAFT_PART_INSTALL}/home"
```
issue: We want the contents of $HOME to go under the `jovyan` user:
```suggestion
      cp -r "${HOME}" "${CRAFT_PART_INSTALL}/home/${NB_USER}"
```

BON4 replied:
It seams to be working in the original way:
```
(base) root@f321d063cf44:/# ls -la /home/jovyan/
total 40
drwxrwsr-x 1 _daemon_ root 4096 Apr 18 12:49 .
drwxr-xr-x 1 root     root 4096 Apr 18 12:43 ..
-rw-r--r-- 1 _daemon_ root   56 Apr 18 12:37 .bashrc
drwxr-sr-x 3 _daemon_ root 4096 Apr 18 12:49 .cache
drwxr-xr-x 2 _daemon_ root 4096 Apr 18 12:43 .conda
drwxr-xr-x 3 _daemon_ root 4096 Apr 18 12:43 .config
drwxr-xr-x 1 _daemon_ root 4096 Apr 18 12:43 .local
(base) root@f321d063cf44:/# ls -la /tmp_home/jovyan/
total 24
drwxrwsr-x 5 _daemon_ root 4096 Apr 18 12:44 .
drwxr-xr-x 3 root     root 4096 Apr 18 12:43 ..
-rw-r--r-- 1 _daemon_ root   56 Apr 18 12:37 .bashrc
drwxr-xr-x 2 _daemon_ root 4096 Apr 18 12:44 .conda
drwxr-xr-x 3 _daemon_ root 4096 Apr 18 12:43 .config
drwxr-xr-x 3 _daemon_ root 4096 Apr 18 12:43 .local
(base) root@f321d063cf44:/# 
```

BON4 commented on codeserver-python/rockcraft.yaml:
```diff
@@ -0,0 +1,246 @@
+    override-build: |
+      echo "en_US.UTF-8 UTF-8" > /etc/locale.gen
+      locale-gen
+      mkdir -p $CRAFT_PART_INSTALL/usr/lib/locale $CRAFT_PART_INSTALL/usr/share/i18n/locales
+      cp /usr/lib/locale/locale-archive $CRAFT_PART_INSTALL/usr/lib/locale/locale-archive
+      cp -r /usr/share/i18n/locales $CRAFT_PART_INSTALL/usr/share/i18n/locales
```
I'm not sure how to test locale. And can't confirm that we do not need this part here. Any suggestins?

BON4 replied:
I assume we're copying all locale runtime artifacts. Since it's unclear which specific locales are used by the Rock packages, the script copies them all by default.

BON4 commented:
@mvlassis I've made some changes based on the [jupyter-tensorflow-full](https://github.com/canonical/kubeflow-rocks/pull/195) Rock review.

Specifically:
- Added support for copying the `tmp_home` directory to `home` at Rock startup, to match the original setup that used `s6` in Docker.  
  See: [discussion](https://github.com/canonical/kubeflow-rocks/pull/195#discussion_r2035468820)

- Removed the user creation step from the overlay script, since it's not being used. The Rock runs with the `_daemon_` user.  
  See: [discussion](https://github.com/canonical/kubeflow-rocks/pull/195#discussion_r2035297899)

mvlassis reviewed (approved):
Great work as always, @BON4!
